File: app/src/main/java/com/rossyn/blocktiles/game2048/presentation/activities/BaseActivity.java
package com.rossyn.blocktiles.game2048.presentation.activities;

import android.app.ActivityManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Color;
import android.media.SoundPool;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.util.DisplayMetrics;
import android.view.Window;
import android.view.WindowManager;
import android.view.WindowMetrics;

import androidx.activity.EdgeToEdge;
import androidx.activity.OnBackPressedCallback;
import androidx.activity.SystemBarStyle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;

import com.rossyn.blocktiles.game2048.R;
import com.rossyn.blocktiles.game2048.data.prefs.SharedPref;
import com.rossyn.blocktiles.game2048.services.Music;

import java.util.List;

import timber.log.Timber;

public abstract class BaseActivity extends AppCompatActivity implements DefaultLifecycleObserver {

    private SoundPool soundPool;
    private int clickSoundId = 0;

    public SharedPref sharedPref;
    Music musicService;
    boolean isBound = false;

    private final ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Music.ServiceBinder binder = (Music.ServiceBinder) service;
            musicService = binder.getService();
            isBound = true;
            musicService.startMusic();
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            isBound = false;
            musicService = null;
        }
    };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        EdgeToEdge.enable(this, SystemBarStyle.auto(Color.TRANSPARENT, Color.TRANSPARENT), SystemBarStyle.light(Color.TRANSPARENT, Color.TRANSPARENT));
        super.onCreate(savedInstanceState);
        getLifecycle().addObserver(this);

        Window window = getWindow();
        WindowInsetsControllerCompat windowInsetsController = WindowCompat.getInsetsController(window, window.getDecorView());
        windowInsetsController.setSystemBarsBehavior(WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
        windowInsetsController.hide(WindowInsetsCompat.Type.systemBars());


        sharedPref = new SharedPref(this);
        soundPool = new SoundPool.Builder().setMaxStreams(5).build();
        clickSoundId = soundPool.load(this, R.raw.click, 1);

        Timber.tag("BaseMonitor").d("onCreate: ");

        bindMusicService();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        getLifecycle().removeObserver(this);
        if (isBound /*&& musicService != null*/) {
            unbindService(serviceConnection);
            isBound = false;
        }
    }


    private void bindMusicService() {
        if (isAppInForeground()) {
            Intent intent = new Intent(this, Music.class);
            bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);

        } else {
            Timber.tag("ServiceError").e("Cannot start service as the app is not in the foreground.");
        }
    }

    /**
     * Called when the activity becomes visible. Registers a callback to handle back press events.
     * If enabled, the back press is intercepted, disabled, and the custom method {@link #onBackPressedCustom()} is called.
     */
    @Override
    protected void onStart() {
        super.onStart();

        Timber.tag("BaseMonitor").d("onStart: ");
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            /**
             * for new Api Level
             * Callback for handling the onBackPressed event.
             */
            @Override
            public void handleOnBackPressed() {
                if (isEnabled()) {
                    setEnabled(true);
                    onBackPressedCustom();
                }
            }
        });
    }

    /**
     * Override this method to provide custom back press behavior.
     */
    protected abstract void onBackPressedCustom();

    public void playClick() {
        if (!sharedPref.getBoolean(SharedPref.MUTE_MUSIC)) {
            soundPool.play(clickSoundId, 1f, 1f, 0, 0, 1f);
        }
    }


    @Override
    public void onResume(@NonNull LifecycleOwner owner) {
        DefaultLifecycleObserver.super.onResume(owner);
        if (musicService != null) musicService.resumeMusic();
        Timber.tag("BaseMonitor").d("onResume: ");
    }

    @Override
    public void onPause(@NonNull LifecycleOwner owner) {
        DefaultLifecycleObserver.super.onPause(owner);
        if (musicService != null) musicService.pauseMusic();
        Timber.tag("BaseMonitor").d("onPause: ");
    }

    @Override
    protected void onUserLeaveHint() {
        super.onUserLeaveHint();
        if (musicService != null) musicService.pauseMusic();
    }

    private boolean isAppInForeground() {
        ActivityManager activityManager = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);
        if (activityManager == null) {
            return false;
        }

        List<ActivityManager.RunningAppProcessInfo> appProcesses = activityManager.getRunningAppProcesses();
        if (appProcesses == null) {
            return false;
        }

        String packageName = getPackageName();
        for (ActivityManager.RunningAppProcessInfo appProcess : appProcesses) {
            if (appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND && appProcess.processName.equals(packageName)) {
                return true;
            }
        }
        return false;
    }


    public void keepScreenOn() {
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    public void keepScreenOff() {
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
    }

    public int getScreenHeight() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            WindowMetrics windowMetrics = getWindowManager().getCurrentWindowMetrics();
            return windowMetrics.getBounds().height();
        } else {
            DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
            return displayMetrics.heightPixels;
        }
    }

    public int getScreenWidth() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            WindowMetrics windowMetrics = getWindowManager().getCurrentWindowMetrics();
            return windowMetrics.getBounds().width();
        } else {
            DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
            return displayMetrics.widthPixels;
        }
    }


    public void navigateTo(Class<?> destination) {
        startActivity(new Intent(this, destination));
    }

}

File: app-kts/src/main/java/com/rossyn/blocktiles/game2048/presentation/activities/BaseActivity.kt
package com.rossyn.blocktiles.game2048.presentation.activities

import android.app.ActivityManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.SoundPool
import android.os.Bundle
import android.os.IBinder
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.lifecycle.DefaultLifecycleObserver
import com.rossyn.blocktiles.game2048.R
import com.rossyn.blocktiles.game2048.data.prefs.SharedPref
import com.rossyn.blocktiles.game2048.services.MusicService
import dagger.hilt.android.AndroidEntryPoint
import timber.log.Timber
import javax.inject.Inject

@AndroidEntryPoint
abstract class BaseActivity : AppCompatActivity(), DefaultLifecycleObserver {

    private var mIsBound = false

    private var mServ: MusicService? = null

    @Inject
    lateinit var sharedPref: SharedPref

    private val serviceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName?, binder: IBinder?) {
            mServ = (binder as MusicService.LocalBinder).getService()
            mIsBound = true
        }

        override fun onServiceDisconnected(name: ComponentName?) {
            mServ = null
            mIsBound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super<AppCompatActivity>.onCreate(savedInstanceState)
        lifecycle.addObserver(this)

        val windowInsetsController = WindowCompat.getInsetsController(window, window.decorView)
        windowInsetsController.systemBarsBehavior =
            WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())


        Timber.tag("BaseMonitor").d("onCreate: ")

        soundPool = SoundPool.Builder()
            .setMaxStreams(5)
            .build()
        clickSoundId = soundPool.load(this, R.raw.click, 1)

        bindMusicService()
    }

    /**
     * Called when the activity becomes visible. Registers a callback to handle back press events.
     * If enabled, the back press is intercepted, disabled, and the custom method {@link #onBackPressedCustom()} is called.
     */
    override fun onStart() {
        super<AppCompatActivity>.onStart()

        Timber.tag("BaseMonitor").d("onStart: ")
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            /**
             * for new Api Level
             * Callback for handling the onBackPressed event.
             */
            override fun handleOnBackPressed() {
                if (isEnabled) {
                    isEnabled = true
                    onBackPressedCustom()
                }
            }
        })

    }

    /**
     * Override this method to provide custom back press behavior.
     */
    abstract fun onBackPressedCustom()


    private fun bindMusicService() {
        if (isAppInForeground()) {
            Intent(this, MusicService::class.java).also {
                startService(it)
                bindService(it, serviceConnection, Context.BIND_AUTO_CREATE)
            }
        } else {
            Timber.tag("ServiceError")
                .e("Cannot start service as the app is not in the foreground.")
        }
    }

    private lateinit var soundPool: SoundPool
    private var clickSoundId: Int = 0

    fun playClick() {
        if (!sharedPref.isMuteMusic) {
            soundPool.play(clickSoundId, 1f, 1f, 0, 0, 1f)
        }
    }


    /**
     * Called when the activity will start interacting with the user.
     * Resets the inactivity timer.
     */
    override fun onResume() {
        super<AppCompatActivity>.onResume()
        mServ?.resumeMusic()
        Timber.tag("BaseMonitor").d("onResume: ")
    }


    /**
     * Called when the system is about to start resuming another activity.
     * Removes any pending inactivity runnable.
     */
    override fun onPause() {
        mServ?.pauseMusic()
        Timber.tag("BaseMonitor").d("onPause: ")
        super<AppCompatActivity>.onPause()
    }


    /**
     * On destroy
     * Remove Lifecycle observer
     *
     */
    override fun onDestroy() {
        lifecycle.removeObserver(this)
        unbindMusicService()
        soundPool.release()
        Timber.tag("BaseMonitor").d("onDestroy")
        super<AppCompatActivity>.onDestroy()
    }

    override fun onUserLeaveHint() {
        super.onUserLeaveHint()
        mServ?.pauseMusic()
    }


    private fun unbindMusicService() {
        if (mIsBound) {
            unbindService(serviceConnection)
            mIsBound = false
        }
    }


    private fun isAppInForeground(): Boolean {
        val activityManager = getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val appProcesses = activityManager.runningAppProcesses ?: return false

        val packageName = packageName
        for (appProcess in appProcesses) {
            if (appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND
                && appProcess.processName == packageName
            ) {
                return true
            }
        }
        return false
    }

}